File: Company Specific Interview Questions/Airbnb/Solutions/Cpp/TwoSum.cpp
//TwoSum
//Given a list of numbers and a number k, return indices of two numbers from the list add up to k.

#include <iostream>
#include <vector>
#include<map>

using namespace std;

vector<int> twoSum(vector<int>& nums, int target) {
    map <int, int> mp;
    vector<int> result;
    int rem;

    for(int i = 0; i < nums.size(); i++) {
        rem = target - nums[i];
        if(mp.find(rem) != mp.end() && mp.find(rem)->second != i) {
            result.push_back(mp.find(rem)->second);
            result.push_back(i);

            return result;
        }
        mp.insert(pair<int, int>(nums[i], i));
    }
    return result;
}

// Input format: n (number of terms in array), target
// 			  	 nums(array of length n)
// Output: Indices of numbers that give the required target sum

int main() {
	int t;
	cin >> t;

	while(t--) {
		vector<int> nums;
		int target, n;
		cin >> n >> target;
		nums.resize(n);
		for(int i = 0; i < n; i++) {
			cin >> nums[i];
		}
		vector<int> result = twoSum(nums, target);

		if(result.size() == 2) {
			cout << result[0] << " " << result[1] << "\n";
		}
		else 
			cout << "No result found" <<"\n";
	}
}

File: Company Specific Interview Questions/Airbnb/Solutions/Python/TwoSum.py
def twoSum(nums, target):
    """
    :type nums: List[int]
    :type target: int
    :rtype: List[int]
    """
    index_map = {}
    for i in range(len(nums)):
        num = nums[i]
        pair = target - num
        if pair in index_map:
            return [index_map[pair], i]
        index_map[num] = i
    return None
 
nums = [1,4,8,3,2,9,15]
target = 13
print("Nums: ", nums)
print("Target: ", target)
print("Solution: ", twoSum(nums, target))
 
print()
nums = [2,7,11,15]
target = 9
print("Nums: ", nums)
print("Target: ", target)
print("Solution: ", twoSum(nums, target))
